### PR TITLE
Fix nested defaults usage

### PR DESCRIPTION
**Pull request by kobik**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -19,8 +18,8 @@ requester.jar = rp.jar;
 requester.cookie = rp.cookie;
 requester.forever = rp.forever;
 
-requester.defaults = (defaultOptions) => {
-    const requester = decorateMethod(defaultOptions);
+function defaults(defaultOptions) {
+    const requester = decorateMethod(undefined, defaultOptions);
```
[the remark this answers is not in the document]
Comment (kobik) [reply index.js]: this sets on the the exported object the default `get` method

File: index.js [hunk 2]
```diff
@@ -32,12 +31,15 @@ requester.defaults = (defaultOptions) => {
     requester.jar = rp.jar;
     requester.cookie = rp.cookie;
     requester.forever = rp.forever;
+    requester.defaults = defaults;
```
[the remark this answers is not in the document]
Comment (kobik) [reply index.js]: Good one, just verified the origin request-promise behavior and takes also the previous ones.

So i corrected the behavior to do the same.

File: index.js [hunk 3]
```diff
@@ -60,12 +62,12 @@ function decorateMethod(method, defaultsOptions = { max: 1 }) {
     };
 }
 
-function buildOptions(uri, method, options = {}, defaultOptions) {
+function buildOptions(uri, method = 'get', options = {}, defaultOptions = {}) {
```
[the remark this answers is not in the document]
Comment (kobik) [reply index.js]: This is exactly the same behavior as request-promise, I'm not sure it's worth mentioning

File: index.js [hunk 4]
```diff
@@ -60,12 +62,12 @@ function decorateMethod(method, defaultsOptions = { max: 1 }) {
...
     if (typeof uri === 'object') {
-        const newOptions = defaults({ method }, uri, defaultOptions, { method: 'get' });
+        const newOptions = Object.assign({ method, max: 1 }, defaultOptions, uri);
         return buildRequestOptions(newOptions);
     } else {
-        const newOptions = defaults({ uri, method }, options, defaultOptions, { method: 'get' });
+        const newOptions = Object.assign({ method, max: 1 }, defaultOptions, options, { uri });
```
[the remark this answers is not in the document]
Comment (kobik) [reply index.js]: correct, fixed the behavior and added more tests

File: index.js [hunk 5]
```diff
@@ -60,12 +62,12 @@ function decorateMethod(method, defaultsOptions = { max: 1 }) {
     };
 }
 
-function buildOptions(uri, method, options = {}, defaultOptions) {
+function buildOptions(uri, method = 'get', options = {}, defaultOptions = {}) {
     if (typeof uri === 'object') {
-        const newOptions = defaults({ method }, uri, defaultOptions, { method: 'get' });
+        const newOptions = Object.assign({ method, max: 1 }, defaultOptions, uri);
```
[the remark this answers is not in the document]
Comment (kobik) [reply index.js]: same as above

File: index.js [hunk 6]
```diff
@@ -32,12 +31,17 @@ requester.defaults = (defaultOptions) => {
     requester.jar = rp.jar;
     requester.cookie = rp.cookie;
     requester.forever = rp.forever;
+    requester.defaults = (newDefaultOptions) => {
+        return defaults(Object.assign({}, defaultOptions, newDefaultOptions));
```
Comment (galcohen92) [inline index.js]: what about all other methods?

**Pull request merged**